### DISCUSSION TITLE
【新提案】获取for/for await的return值

### DISCUSSION THREAD

**Discussion by LongTengDao** [JSCIG/es-discuss#51] [Ideas]
```
function * g () { return 1; }
const v = for ( const each of g() ) {}
console.log(v);// 1
```

时不时需要获知return值，目前就得额外参数通信，或者改成手动调用next，很不方便。

**Comment by hax** [comment]
去年 iterator helpers 提案的讨论决定 iterator helpers 不考虑 generator 协议在单纯 iterator 协议之外的额外语义，如 `next(v)` 传入的值，而generator的return值也是这种情况。所以很遗憾，这个提议不太可能通过委员会。

另一个问题是，for-of（以及所有的循环结构）虽然并不是表达式，但作为语句也是有值的，可以通过 eval 来得到。其语义为循环体中最后一个语句的值（如果没有则为`undefined`）。

```js
const v = eval(` for (const each of [1, 2]) { 42 } `)
v // 42
```

因此如果要允许 for-of 成为表达式，其值的语义恐怕也很难改变。
